### ISSUE TITLE
Circular references

### ISSUE THREAD

**Issue by nicolas-t** [justinrainbow/json-schema#281]
Hi,

I'm using the v2.0.5, and I encouter this error:

`PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 32 bytes) in vendor/justinrainbow/json-schema/src/JsonSchema/Iterator/ObjectIterator.php on line 127`

This error seems to be related to circular references. 

Here is a minimal version of my 2 schemas:

`Author.json ` ***in some cases*** will return a collection of Posts
```json
{
    "$schema": "http://json-schema.org/draft-04/schema",
    "id": "Author.json",
    "type": "object",
    "properties": {
        ...
        "posts": {
            "type": "array",
            "items": {
                "$ref": "Post.json#"
            }
        }
    }
}
```

`Post.json` ***in some cases*** will return an Author
```json
{
    "$schema": "http://json-schema.org/draft-04/schema",
    "id": "Post.json",
    "title": "Post Schema",
    "type": "object",
    "properties": {
        ...
        "author": {
            "$ref": "Author.json#"
        }
    }
}
```
I just have two cases:

1. I want an author with a collection of posts

2. I want a post with his author

The circular cases (where I want an author with his posts+author) never happens in real life.

Not sure if i'm clear.

My question:
Is it a problem of the way I designed my schemas or is it a problem of your validation library?

Best regards,

**Comment by csimplestring** [comment]
@nicolas-t Can you check this issue #274 ? In my comment, I found that

In our code, we want to cache all the resolved schemas, so we write like this:
```
// This will cause infinite looping and crash
$refResolver = new RefResolver(new UriRretriever(), new UriResolver());
for ($schemas as $schema) {
  $resolved = $refResolver->resolver($schema);
  $cache[] = $resolved;
}

// but this will work ok
for ($schemas as $schema) {
  $refResolver = new RefResolver(new UriRretriever(), new UriResolver());
  $resolved = $refResolver->resolver($schema);
  $cache[] = $resolved;
}
```

It seems that every time you dereference a schema, you need to create a new RefResolver.

**Comment by nicolas-t** [comment]
Thanks !

yes I saw your comment before opening this issue, but in my code i'm not looping over schemas.

```php
$refResolver = new JsonSchema\RefResolver(new JsonSchema\Uri\UriRetriever(), new JsonSchema\Uri\UriResolver());
$schema = $refResolver->resolve('file://mySchema.json'));

$validator = new JsonSchema\Validator();
$validator->check($json->data, $schema);
```

So i'm not sure how to use your example.

**Comment by bighappyface** [comment]
Should be solved with #277 and is available in 3.0.0
